tests/test_BT_piecewise.py

```diff
@@ -0,0 +1,560 @@
+)
+# assert array equal truth_array not (~) in_piece
+def test_d_delay_in_groups_respond_to_changes_in_corresponding_piecewise_parameter(
+    build_model, build_piecewise_model_with_two_pieces, param, index
+):
+    m_piecewise, toas = add_groups_and_make_toas(
+        build_model, build_piecewise_model_with_two_pieces, "overlapping groups"
+    )
+    m_piecewise_temp = add_offset_in_model_parameter(index, param, m_piecewise)
+    parameter_string = f"{param}_{convert_int_into_index(index)}"
+    if index == -1:
+        is_d_delay_changing = np.invert(
+            np.isclose(
+                m_piecewise_temp.d_binary_delay_d_xxxx(toas, param[0:2], None).value,
+                0,
+                atol=1e-11,
+                rtol=0,
+            )
+        )
+    else:
+        is_d_delay_changing = np.invert(
+            np.isclose(
+                m_piecewise_temp.d_binary_delay_d_xxxx(
+                    toas, parameter_string, None
+                ).value,
+                0,
+                atol=1e-11,
+                rtol=0,
+            )
+        )
+    should_d_delay_be_changing = m_piecewise.does_toa_reference_piecewise_parameter(
+        toas, parameter_string
+    )
+    # assert toas that are in the group/gap have some non-zero delay derivative
+    both_arrays_should_be_equal = np.array_equal(
+        is_d_delay_changing, should_d_delay_be_changing
+    )
+    assert both_arrays_should_be_equal
+
+
+def add_relative_offset_for_derivatives(index, param, model, offset_size, plus=True):
+    m_piecewise_temp = deepcopy(model)
+    parameter_string = f"{param}_{convert_int_into_index(index)}"
+    offset_size = offset_size.value
+    if plus is True:
+        if hasattr(m_piecewise_temp, parameter_string):
+            delta = getattr(m_piecewise_temp, parameter_string).value + offset_size
+            getattr(m_piecewise_temp, parameter_string).value = delta
+            m_piecewise_temp.setup()
+        else:
+            parameter_string = param[0:2]
+            getattr(m_piecewise_temp, parameter_string).value = (
+                getattr(m_piecewise_temp, parameter_string).value + offset_size
+            )
+            m_piecewise_temp.setup()
+        return m_piecewise_temp
+    else:
+        if hasattr(m_piecewise_temp, parameter_string):
+            delta = getattr(m_piecewise_temp, parameter_string).value - offset_size
+            getattr(m_piecewise_temp, parameter_string).value = delta
+            m_piecewise_temp.setup()
+        else:
+            parameter_string = param[0:2]
+            getattr(m_piecewise_temp, parameter_string).value = (
+                getattr(m_piecewise_temp, parameter_string).value - offset_size
+            )
+            m_piecewise_temp.setup()
+        return m_piecewise_temp
+
+
+def get_d_delay_d_xxxx(toas, model, parameter_string):
+    rs_temp = pint.residuals.Residuals(toas, model, subtract_mean=False)
```

Comment:
Do you really need to construct Residuals here? Why?

Reply:
It appears given the updates I don't (/perhaps never needed to), now constructing residuals is reserved for tests that actually use them.